File: TCP Split File/SplitFileClient.go
/* 20195914 */
/* Jeong eui chan */

package main

import (
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"strings"
)

const (
	server1Address = "nsl2.cau.ac.kr:45914"
	server2Address = "nsl5.cau.ac.kr:55914"
)

func main() {
	// Check for the correct number of command-line arguments
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run SplitFileClient.go <put/get> <filename>")
		return
	}

	command := os.Args[1]
	filename := os.Args[2]

	// Execute the appropriate function based on the command
	switch command {
	case "put":
		putFile(filename)
	case "get":
		getFile(filename)
	default:
		fmt.Println("Unknown command:", command)
	}
}

// Function to handle the 'put' command
func putFile(filename string) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	// Split the file data into two parts
	part1, part2 := splitFile(data)
	part1Filename := strings.TrimSuffix(filename, ".txt") + "-part1.txt"
	part2Filename := strings.TrimSuffix(filename, ".txt") + "-part2.txt"

	// Send each part to the respective server
	if !sendToServer(server1Address, "put "+part1Filename, part1) {
		fmt.Println("Failed to send part1 to server 1")
		return
	}
	if !sendToServer(server2Address, "put "+part2Filename, part2) {
		fmt.Println("Failed to send part2 to server 2")
		return
	}
}

// Function to handle the 'get' command
func getFile(filename string) {
	part1Filename := strings.TrimSuffix(filename, ".txt") + "-part1.txt"
	part2Filename := strings.TrimSuffix(filename, ".txt") + "-part2.txt"

	// Receive each part from the respective server
	part1 := receiveFromServer(server1Address, "get "+part1Filename)
	if part1 == nil {
		fmt.Println("Failed to receive part1 from server 1")
		return
	}

	part2 := receiveFromServer(server2Address, "get "+part2Filename)
	if part2 == nil {
		fmt.Println("Failed to receive part2 from server 2")
		return
	}

	// Merge the two parts and save to a new file
	merged := mergeFile(part1, part2)
	mergedFilename := strings.TrimSuffix(filename, ".txt") + "-merged.txt"
	err := ioutil.WriteFile(mergedFilename, merged, 0644)
	if err != nil {
		fmt.Println("Error writing merged file:", err)
		return
	}
	fmt.Println("Merged file saved as", mergedFilename)
}

// Function to split the data into two parts
func splitFile(data []byte) ([]byte, []byte) {
	var part1, part2 []byte
	for i := 0; i < len(data); i++ {
		if i%2 == 0 {
			part1 = append(part1, data[i])
		} else {
			part2 = append(part2, data[i])
		}
	}
	return part1, part2
}

// Function to merge two parts into one
func mergeFile(part1, part2 []byte) []byte {
	var merged []byte
	i, j := 0, 0
	for i < len(part1) || j < len(part2) {
		if i < len(part1) {
			merged = append(merged, part1[i])
			i++
		}
		if j < len(part2) {
			merged = append(merged, part2[j])
			j++
		}
	}
	return merged
}

// Function to send data to a server
func sendToServer(address, command string, data []byte) bool {
	conn, err := net.Dial("tcp", address)
	if err != nil {
		fmt.Println("Error connecting to server:", err)
		return false
	}
	defer conn.Close()

	// Send the command to the server
	_, err = fmt.Fprintln(conn, command)
	if err != nil {
		fmt.Println("Error sending command:", err)
		return false
	}

	// Send the data to the server
	_, err = conn.Write(data)
	if err != nil {
		fmt.Println("Error sending data:", err)
		return false
	}
	return true
}

// Function to receive data from a server
func receiveFromServer(address, command string) []byte {
	conn, err := net.Dial("tcp", address)
	if err != nil {
		fmt.Println("Error connecting to server:", err)
		return nil
	}
	defer conn.Close()

	// Send the command to the server
	_, err = fmt.Fprintln(conn, command)
	if err != nil {
		fmt.Println("Error sending command:", err)
		return nil
	}

	// Read the response from the server
	data, err := ioutil.ReadAll(conn)
	if err != nil {
		fmt.Println("Error receiving data:", err)
		return nil
	}

	// Check for an error message from the server
	if strings.HasPrefix(string(data), "Error:") {
		fmt.Println(string(data))
		return nil
	}

	return data
}

File: TCP_Split_File/SplitFileClient_rust/src/SplitFileClient.rs
/* 20195914 */
/* Jeong eui chan */

use std::env;
use std::fs::File;
use std::io::{Read, Write, BufReader, BufWriter};
use std::net::TcpStream;

const SERVER1_ADDRESS: &str = "nsl2.cau.ac.kr:45914";
const SERVER2_ADDRESS: &str = "nsl5.cau.ac.kr:55914";

// Function to send data to the server
fn send_to_server(address: &str, command: &str, data: &[u8]) -> bool {
    // Connect to the server
    let stream = match TcpStream::connect(address) {
        Ok(stream) => stream,
        Err(e) => {
            eprintln!("Error connecting to server: {}", e);
            return false;
        }
    };

    let mut writer = BufWriter::new(&stream);
    // Send the command to the server
    if writeln!(writer, "{}", command).is_err() {
        eprintln!("Error sending command");
        return false;
    }

    // Send the data to the server
    if writer.write_all(data).is_err() {
        eprintln!("Error sending data");
        return false;
    }

    true
}

// Function to receive data from the server
fn receive_from_server(address: &str, command: &str) -> Vec<u8> {
    // Connect to the server
    let stream = TcpStream::connect(address).expect("Error connecting to server");
    let mut writer = BufWriter::new(&stream);
    writeln!(writer, "{}", command).expect("Error sending command");

    let mut reader = BufReader::new(&stream);
    let mut buffer = Vec::new();
    // Read the data from the server
    reader.read_to_end(&mut buffer).expect("Error receiving data");

    // Check for error message from the server
    if buffer.starts_with(b"Error:") {
        eprintln!("{}", String::from_utf8_lossy(&buffer));
        return Vec::new();
    }

    buffer
}

// Function to handle the 'put' command
fn put_file(filename: &str) {
    let mut file = File::open(filename).expect("Error reading file");
    let mut data = Vec::new();
    file.read_to_end(&mut data).expect("Error reading file");

    let (part1, part2) = split_file(&data);

    let part1_filename = format!("{}-part1.txt", filename.trim_end_matches(".txt"));
    let part2_filename = format!("{}-part2.txt", filename.trim_end_matches(".txt"));

    // Send each part to the respective server
    if !send_to_server(SERVER1_ADDRESS, &format!("put {}", part1_filename), &part1) {
        eprintln!("Failed to send part1 to server 1");
        return;
    }

    if !send_to_server(SERVER2_ADDRESS, &format!("put {}", part2_filename), &part2) {
        eprintln!("Failed to send part2 to server 2");
    }
}

// Function to handle the 'get' command
fn get_file(filename: &str) {
    let part1_filename = format!("{}-part1.txt", filename.trim_end_matches(".txt"));
    let part2_filename = format!("{}-part2.txt", filename.trim_end_matches(".txt"));

    // Receive each part from the respective server
    let part1 = receive_from_server(SERVER1_ADDRESS, &format!("get {}", part1_filename));
    if part1.is_empty() {
        eprintln!("Failed to receive part1 from server 1");
        return;
    }

    let part2 = receive_from_server(SERVER2_ADDRESS, &format!("get {}", part2_filename));
    if part2.is_empty() {
        eprintln!("Failed to receive part2 from server 2");
        return;
    }

    // Merge the two parts and save to a new file
    let merged = merge_file(&part1, &part2);
    let merged_filename = format!("{}-merged.txt", filename.trim_end_matches(".txt"));
    let mut file = File::create(merged_filename).expect("Error writing merged file");
    file.write_all(&merged).expect("Error writing merged file");
    println!("Merged file saved as {}-merged.txt", filename.trim_end_matches(".txt"));
}

// Function to split the data into two parts
fn split_file(data: &[u8]) -> (Vec<u8>, Vec<u8>) {
    let mut part1 = Vec::new();
    let mut part2 = Vec::new();
    for (i, &byte) in data.iter().enumerate() {
        if i % 2 == 0 {
            part1.push(byte);
        } else {
            part2.push(byte);
        }
    }
    (part1, part2)
}

// Function to merge two parts into one
fn merge_file(part1: &[u8], part2: &[u8]) -> Vec<u8> {
    let mut merged = Vec::new();
    let mut i = 0;
    let mut j = 0;
    while i < part1.len() || j < part2.len() {
        if i < part1.len() {
            merged.push(part1[i]);
            i += 1;
        }
        if j < part2.len() {
            merged.push(part2[j]);
            j += 1;
        }
    }
    merged
}

// Main function to parse command-line arguments and execute the appropriate command
fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: cargo run -- <put/get> <filename>");
        return;
    }

    let command = &args[1];
    let filename = &args[2];

    match command.as_str() {
        "put" => put_file(filename),
        "get" => get_file(filename),
        _ => eprintln!("Unknown command: {}", command),
    }
}
